# Data Preparation


Unduh Data CSV dari https://github.com/realpython/python-data-cleaning/blob/master/Datasets/BL-Flickr-Images-Book.csv

### 1. Import Library yang dibutuhkan

In [1]:
import pandas as pd
import numpy as np



Membaca Data

In [21]:
df = pd.read_csv("BL-FLickr-Images-Book.csv")

Membaca secara cepat

In [22]:
df.describe()

Identifier  Corporate Author  Corporate Contributors  Engraver
count  8.287000e+03               0.0                     0.0       0.0
mean   2.017344e+06               NaN                     NaN       NaN
std    1.190379e+06               NaN                     NaN       NaN
min    2.060000e+02               NaN                     NaN       NaN
25%    9.157875e+05               NaN                     NaN       NaN
50%    2.043707e+06               NaN                     NaN       NaN
75%    3.047430e+06               NaN                     NaN       NaN
max    4.160339e+06               NaN                     NaN       NaN

In [23]:
df

Identifier             Edition Statement      Place of Publication  \
0            206                           NaN                    London   
1            216                           NaN  London; Virtue & Yorston   
2            218                           NaN                    London   
3            472                           NaN                    London   
4            480  A new edition, revised, etc.                    London   
...          ...                           ...                       ...   
8282     4158088                           NaN                    London   
8283     4158128                           NaN                     Derby   
8284     4159563                           NaN                    London   
8285     4159587                           NaN       Newcastle upon Tyne   
8286     4160339                           NaN                    London   

     Date of Publication                Publisher  \
0            1879 [1878]         S. Tinsley & Co.   
1                   1868             Virtue & Co.   
2                   1869    Bradbury, Evans & Co.   
3                   1851            James Darling   
4                   1857     Wertheim & Macintosh   
...                  ...                      ...   
8282                1838                      NaN   
8283            1831, 32          M. Mozley & Son   
8284           [1806]-22  T. Cadell and W. Davies   
8285                1834         Mackenzie & Dent   
8286             1834-43                      NaN   

                                                  Title  \
0                     Walter Forbes. [A novel.] By A. A   
1     All for Greed. [A novel. The dedication signed...   
2     Love the Avenger. By the author of “All for Gr...   
3     Welsh Sketches, chiefly ecclesiastical, to the...   
4     [The World in which I live, and my place in it...   
...                                                 ...   
8282  The Parochial History of Cornwall, founded on,...   
8283  The History and Gazetteer of the County of Der...   
8284  Magna Britannia; being a concise topographical...   
8285  An historical, topographical and descriptive v...   
8286  Collectanea Topographica et Genealogica. [Firs...   

                                                 Author  \
0                                                 A. A.   
1                                             A., A. A.   
2                                             A., A. A.   
3                                             A., E. S.   
4                                             A., E. S.   
...                                                 ...   
8282                 GIDDY, afterwards GILBERT, Davies.   
8283                         GLOVER, Stephen - of Derby   
8284  LYSONS, Daniel - M.A., F.R.S., and LYSONS (Sam...   
8285                              Mackenzie, E. (Eneas)   
8286                                                NaN   

                                           Contributors  Corporate Author  \
0                                       FORBES, Walter.               NaN   
1          BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
2          BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
3                           Appleyard, Ernest Silvanus.               NaN   
4                                   BROOME, John Henry.               NaN   
...                                                 ...               ...   
8282  BOASE, Henry Samuel.|HALS, William.|LYSONS, Da...               NaN   
8283                                     NOBLE, Thomas.               NaN   
8284           GREGSON, Matthew.|LYSONS, Samuel - F.R.S               NaN   
8285                               ROSS, M. - of Durham               NaN   
8286            BANDINEL, Bulkeley.|Nichols, John Gough               NaN   

      Corporate Contributors Former owner  Engraver Issuance type  \
0                        NaN          NaN   

Menghapus beberapa Kolom

In [24]:
deleteList2 = ['Edition Statement', 'Corporate Author','Corporate Contributors', 'Former owner', 'Engraver', 'Contributors','Issuance type', 'Shelfmarks' ]
df.drop(deleteList2, inplace=True, axis=1)

In [25]:
df

Identifier      Place of Publication Date of Publication  \
0            206                    London         1879 [1878]   
1            216  London; Virtue & Yorston                1868   
2            218                    London                1869   
3            472                    London                1851   
4            480                    London                1857   
...          ...                       ...                 ...   
8282     4158088                    London                1838   
8283     4158128                     Derby            1831, 32   
8284     4159563                    London           [1806]-22   
8285     4159587       Newcastle upon Tyne                1834   
8286     4160339                    London             1834-43   

                    Publisher  \
0            S. Tinsley & Co.   
1                Virtue & Co.   
2       Bradbury, Evans & Co.   
3               James Darling   
4        Wertheim & Macintosh   
...                       ...   
8282                      NaN   
8283          M. Mozley & Son   
8284  T. Cadell and W. Davies   
8285         Mackenzie & Dent   
8286                      NaN   

                                                  Title  \
0                     Walter Forbes. [A novel.] By A. A   
1     All for Greed. [A novel. The dedication signed...   
2     Love the Avenger. By the author of “All for Gr...   
3     Welsh Sketches, chiefly ecclesiastical, to the...   
4     [The World in which I live, and my place in it...   
...                                                 ...   
8282  The Parochial History of Cornwall, founded on,...   
8283  The History and Gazetteer of the County of Der...   
8284  Magna Britannia; being a concise topographical...   
8285  An historical, topographical and descriptive v...   
8286  Collectanea Topographica et Genealogica. [Firs...   

                                                 Author  \
0                                                 A. A.   
1                                             A., A. A.   
2                                             A., A. A.   
3                                             A., E. S.   
4                                             A., E. S.   
...                                                 ...   
8282                 GIDDY, afterwards GILBERT, Davies.   
8283                         GLOVER, Stephen - of Derby   
8284  LYSONS, Daniel - M.A., F.R.S., and LYSONS (Sam...   
8285                              Mackenzie, E. (Eneas)   
8286                                                NaN   

                                             Flickr URL  
0     http://www.flickr.com/photos/britishlibrary/ta...  
1     http://www.flickr.com/photos/britishlibrary/ta...  
2     http://www.flickr.com/photos/britishlibrary/ta...  
3     http://www.flickr.com/photos/britishlibrary/ta...  
4     http://www.flickr.com/photos/britishlibrary/ta...  
...                                                 ...  
8282  http://www.flickr.com/photos/britishlibrary/ta...  
8283  http://www.flickr.com/photos/britishlibrary/ta...  
8284  http://www.flickr.com/photos/britishlibrary/ta...  
8285  http://www.flickr.com/photos/britishlibrary/ta...  
8286  http://www.flickr.com/photos/britishlibrary/ta...  

[8287 rows x 7 columns]

### 2. Mengubah Indeks di Data 

In [26]:
df['Identifier'].is_unique

True

In [27]:
df = df.set_index('Identifier')

In [28]:
df

Place of Publication Date of Publication  \
Identifier                                                 
206                           London         1879 [1878]   
216         London; Virtue & Yorston                1868   
218                           London                1869   
472                           London                1851   
480                           London                1857   
...                              ...                 ...   
4158088                       London                1838   
4158128                        Derby            1831, 32   
4159563                       London           [1806]-22   
4159587          Newcastle upon Tyne                1834   
4160339                       London             1834-43   

                          Publisher  \
Identifier                            
206                S. Tinsley & Co.   
216                    Virtue & Co.   
218           Bradbury, Evans & Co.   
472                   James Darling   
480            Wertheim & Macintosh   
...                             ...   
4158088                         NaN   
4158128             M. Mozley & Son   
4159563     T. Cadell and W. Davies   
4159587            Mackenzie & Dent   
4160339                         NaN   

                                                        Title  \
Identifier                                                      
206                         Walter Forbes. [A novel.] By A. A   
216         All for Greed. [A novel. The dedication signed...   
218         Love the Avenger. By the author of “All for Gr...   
472         Welsh Sketches, chiefly ecclesiastical, to the...   
480         [The World in which I live, and my place in it...   
...                                                       ...   
4158088     The Parochial History of Cornwall, founded on,...   
4158128     The History and Gazetteer of the County of Der...   
4159563     Magna Britannia; being a concise topographical...   
4159587     An historical, topographical and descriptive v...   
4160339     Collectanea Topographica et Genealogica. [Firs...   

                                                       Author  \
Identifier                                                      
206                                                     A. A.   
216                                                 A., A. A.   
218                                                 A., A. A.   
472                                                 A., E. S.   
480                                                 A., E. S.   
...                                                       ...   
4158088                    GIDDY, afterwards GILBERT, Davies.   
4158128                            GLOVER, Stephen - of Derby   
4159563     LYSONS, Daniel - M.A., F.R.S., and LYSONS (Sam...   
4159587                                 Mackenzie, E. (Eneas)   
4160339                                                   NaN   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...  
...                                                       ...  
4158088     http://www.flickr.com/photos/britishlibrary/ta...  
4158128     http://www.flickr.com/photos/britishlibrary/ta...  
4159563     http://www.flickr.com/photos/britishlibrary/ta...  
4159587     http://www.flickr.com/photos/britishlibrary/ta...  
4160339     http://www.flickr.com/photos/britishlibrary/ta...  

[8287 rows x 6 columns]

In [29]:
df.head()

Place of Publication Date of Publication  \
Identifier                                                 
206                           London         1879 [1878]   
216         London; Virtue & Yorston                1868   
218                           London                1869   
472                           London                1851   
480                           London                1857   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...

Mencari Data menggunakan loc

In [31]:
df.loc[472]

Place of Publication                                               London
Date of Publication                                                  1851
Publisher                                                   James Darling
Title                   Welsh Sketches, chiefly ecclesiastical, to the...
Author                                                          A., E. S.
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 472, dtype: object

### 3. Merapikan Fields

Menghitung Kolom

In [33]:
df.dtypes.value_counts()

object    6
dtype: int64

Mencari Data

In [40]:
df.loc[1866:,'Date of Publication'].head(15)

Identifier
1905                  1888
1929           1839, 38-54
2836                  1897
2854                  1865
2956               1860-63
2957                  1873
3017                  1866
3131                  1899
4598                  1814
4884                  1820
4976                  1800
5382    1847, 48 [1846-48]
5385               [1897?]
5389               [1897?]
5432                  1893
Name: Date of Publication, dtype: object

> - Menghiangkan tanggal lain dalam kurung sikut
> - Konversi Rentang Tanggal
> - Hilangkan Tanggal yang gak jelas [1897?] -> NaN
> - Konversi NaN

In [41]:
regex = r'^(\d{4})'

In [54]:
ekstrak = df['Date of Publication'].str.extract(r'^(\d{4})',expand=False)

In [55]:
ekstrak.head(10)

Identifier
206     1879
216     1868
218     1869
472     1851
480     1857
481     1875
519     1872
667      NaN
874     1676
1143    1679
Name: Date of Publication, dtype: object

In [57]:
df['Date of Publication'] = pd.to_numeric(ekstrak)
df['Date of Publication'].dtypes

dtype('float64')

### 4. Membersihkan Kolom

In [58]:
df['Place of Publication'].head(15)

Identifier
206                                  London
216                London; Virtue & Yorston
218                                  London
472                                  London
480                                  London
481                                  London
519                                  London
667     pp. 40. G. Bryan & Co: Oxford, 1898
874                                 London]
1143                                 London
1280                               Coventry
1808                            Christiania
1905                                Firenze
1929                              Amsterdam
2836                                 Savona
Name: Place of Publication, dtype: object

In [59]:
df.loc[667]

Place of Publication                  pp. 40. G. Bryan & Co: Oxford, 1898
Date of Publication                                                   NaN
Publisher                                                             NaN
Title                   The Coming of Spring, and other poems. By J. A...
Author                                                      A., J.|A., J.
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 667, dtype: object

In [62]:
df.loc[481]

Place of Publication                                               London
Date of Publication                                                1875.0
Publisher                                               William Macintosh
Title                   [The World in which I live, and my place in it...
Author                                                          A., E. S.
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 481, dtype: object

Merapikan data String

In [66]:
publication = df['Place of Publication']
london = publication.str.contains('London')
london[:5]

oxford = publication.str.contains('Oxford')

In [67]:
df['Place of Publication'] = np.where(london, 'London', np.where(oxford, 'Oxford', publication.str.replace('-', ' ')))

In [68]:
df.tail(15)

Place of Publication  Date of Publication                Publisher  \
Identifier                                                                      
4115138                  London                  NaN                      NaN   
4116063                  London               1866.0             [John Dicks]   
4117526                 Leipzig               1862.0    Charles Fr. Fleischer   
4117583                  London               1894.0          T. Fisher Unwin   
4117749                New York               1868.0         John Wiley & Son   
4117751                New York               1882.0        John Wiley & Sons   
4117752                  London               1883.0           Trübner & Co.   
4156359                New York               1898.0      G. P. Putnam's Sons   
4157746                  London               1811.0   F. C. and J. Rivington   
4157862     Newcastle upon Tyne               1867.0               T. Fordyce   
4158088                  London               1838.0                      NaN   
4158128                   Derby               1831.0          M. Mozley & Son   
4159563                  London                  NaN  T. Cadell and W. Davies   
4159587     Newcastle upon Tyne               1834.0         Mackenzie & Dent   
4160339                  London               1834.0                      NaN   

                                                        Title  \
Identifier                                                      
4115138     The Revenger's Tagaedie. As it hath beene sund...   
4116063     Edith Heron; or, the Earl and the countess. [B...   
4117526     Abridgment of the History of England ... Adapt...   
4117583     The Wish. A novel by Hermann Sudermann. Transl...   
4117749     A System of Mineralogy. Descriptive Mineralogy...   
4117751     Third Appendix to the fifth edition of Dana's ...   
4117752     A System of Mineralogy ... Fifth edition, rewr...   
4156359     History of the People of the Netherlands ... T...   
4157746     The New Chronicles of England and France, in t...   
4157862     Local Records; or, Historical Register of rema...   
4158088     The Parochial History of Cornwall, founded on,...   
4158128     The History and Gazetteer of the County of Der...   
4159563     Magna Britannia; being a concise topographical...   
4159587     An historical, topographical and descriptive v...   
4160339     Collectanea Topographica et Genealogica. [Firs...   

                                                       Author  \
Identifier                                                      
4115138                                                   NaN   
4116063                                         HERON, Edith.   
4117526                                     Goldsmith, Oliver   
4117583                                   SUDERMANN, Hermann.   
4117749                                   DANA, James Dwight.   
4117751                                   DANA, James Dwight.   
4117752                                   DANA, James Dwight.   
4156359                                   BLOK, Pieter Johan.   
4157746                                       FABYAN, Robert.   
4157862         FORDYCE, T. - Printer, of Newcastle-upon-Tyne   
4158088                    GIDDY, afterwards GILBERT, Davies.   
4158128                            GLOVER, Stephen - of Derby   
4159563     LYSONS, Daniel - M.A., F.R.S., and LYSONS (Sam...   
4159587                                 Mackenzie, E. (Eneas)   
4160339                                                   NaN   

                                                   Flickr URL  
Identifier                                                     
4115138     http://www.flickr.com/photos/britishlibrary/ta...  
4116063     http://www.flickr.com/photos/britishlibrary/ta...  
4117526     http://www.flickr.com/photos/britishlibrary/ta...  
4117583     http://www.flickr.com/photos/britishlibrary/ta...  
4117749     http://www.flickr.com/pho

In [69]:
df.head()

Place of Publication  Date of Publication              Publisher  \
Identifier                                                                    
206                      London               1879.0       S. Tinsley & Co.   
216                      London               1868.0           Virtue & Co.   
218                      London               1869.0  Bradbury, Evans & Co.   
472                      London               1851.0          James Darling   
480                      London               1857.0   Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...

### Dataset Baru
Unduh di https://github.com/realpython/python-data-cleaning/blob/master/Datasets/university_towns.txt

### 5. Membersihkan Dataset dengan ApplyMap

In [72]:
university_town = []
with open('university_towns.txt') as file:
    for line in file:
        if '[edit]' in line:
            state = line
        else :
            university_town.append((state,line))

university_town[:5]

[('Alabama[edit]\n', 'Auburn (Auburn University)[1]\n'),
 ('Alabama[edit]\n', 'Florence (University of North Alabama)\n'),
 ('Alabama[edit]\n', 'Jacksonville (Jacksonville State University)[2]\n'),
 ('Alabama[edit]\n', 'Livingston (University of West Alabama)[2]\n'),
 ('Alabama[edit]\n', 'Montevallo (University of Montevallo)[2]\n')]

In [75]:
dfCity = pd.DataFrame(university_town, columns=['State', 'Region Name'])
dfCity.head()

State                                        Region Name
0  Alabama[edit]\n                    Auburn (Auburn University)[1]\n
1  Alabama[edit]\n           Florence (University of North Alabama)\n
2  Alabama[edit]\n  Jacksonville (Jacksonville State University)[2]\n
3  Alabama[edit]\n       Livingston (University of West Alabama)[2]\n
4  Alabama[edit]\n         Montevallo (University of Montevallo)[2]\n

In [77]:
def getCity(item):
    if '(' in item:
        return item[:item.find('(')]
    elif '[' in item:
        return item[:item.find('[')]
    else:
        return item



In [82]:
dfCity =  dfCity.applymap(getCity)

In [84]:
dfCity.head()

State    Region Name
0  Alabama        Auburn 
1  Alabama      Florence 
2  Alabama  Jacksonville 
3  Alabama    Livingston 
4  Alabama    Montevallo